d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java

```diff
@@ -257,21 +257,25 @@ private void readyHandler()
       _log.warn("Unexpected state, ready called on null or backed off ADS stream!");
       return;
     }
-    // confirm ready state to neglect spurious callbacks; we'll get another callback whenever it is ready again.
-    if (_adsStream.isReady())
+    // Confirm ready state to neglect spurious callbacks; we'll get another callback whenever it is ready again.
+    // Also confirm ready timeout future is not null to avoid notifying multiple times.
+    if (!_adsStream.isReady() || _readyTimeoutFuture == null)
     {
-      // if the ready timeout future is non-null, a reconnect notification hasn't been sent yet.
-      if (_readyTimeoutFuture != null)
-      {
-        // timeout task will be cancelled only if it hasn't already executed.
-        boolean cancelledTimeout = _readyTimeoutFuture.cancel(false);
-        _log.info("ADS stream ready, cancelled timeout task: {}", cancelledTimeout);
-        _readyTimeoutFuture = null; // set it to null to avoid repeat notifications to subscribers.
-        _xdsClientJmx.incrementReconnectionCount();
-        notifyStreamReconnect();
-      }
-      _xdsClientJmx.setIsConnected(true);
+      return;
+    }
+
+    // timeout task will be cancelled only if it hasn't already executed.
+    boolean cancelledTimeout = _readyTimeoutFuture.cancel(false);
+    _log.info("ADS stream ready, cancelled timeout task: {}", cancelledTimeout);
+    _readyTimeoutFuture = null; // set it to null to avoid repeat notifications to subscribers.
```

Comment:
Although not too bad, this might be some unneeded refactoring? (don't fix what ain't broken)

Reply:
yeah, it's just return-early, as suggested by @brycezhongqing above.